engine/src/components/armor.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * armor.cpp
+ *
+ * Copyright (c) 2001-2002 Daniel Horn
+ * Copyright (c) 2002-2019 pyramid3d and other Vega Strike Contributors
+ * Copyright (c) 2019-2024 Stephen G. Tuggy, Benjamen R. Meyer, Roy Falk and other Vega Strike Contributors
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+#include "armor.h"
+
+#include "damageable_layer.h"
+#include "unit_csv_factory.h"
+
+const std::string armor_facets[] = {
+    "Moment_Of_Inertia",
+    "Armor_Front_Top_Right",
+    "Armor_Front_Top_Left",
+    "Armor_Front_Bottom_Right",
+    "Armor_Front_Bottom_Left",
+    "Armor_Back_Top_Right",
+    "Armor_Back_Top_Left",
+    "Armor_Back_Bottom_Right",
+    "Armor_Back_Bottom_Left"
+};
+
+Armor::Armor(): Component("", 0.0, 0.0, false),
+                DamageableLayer(1, FacetConfiguration::eight, 
+                                Health(0,0), false) {}
+
+void Armor::Load(std::string upgrade_key, std::string unit_key) {
+    // Component
+    Component::Load(upgrade_key, unit_key);
+
+ 
+    // Damageable Layer
+    std::string upgrade_type_string = UnitCSVFactory::GetVariable(upgrade_key, "Name", std::string());
+
+    std::vector<double> armor_values;
+
+    // Max values from Upgrade
+    for(int i = 0;i < 8;i++) {
+        double armor_facet_max = UnitCSVFactory::GetVariable(upgrade_key, armor_facets[i], 0.0);
+        armor_values.push_back(armor_facet_max);
+    }
+
+    UpdateFacets(armor_values);
+
+    // Current values from Save Game
+    for(int i = 0;i < 8;i++) {
+        double armor_value = UnitCSVFactory::GetVariable(unit_key, 
+                                               armor_facets[i], 0.0);
+        facets[i].health.Set(armor_value);
+    }
+}
+
+void Armor::SaveToCSV(std::map<std::string, std::string>& unit) const {
+    for(int i=0;i<8;i++) {
+        unit[armor_facets[i]] = std::to_string(facets[i].health.Value());
+    }
```

Comment:
is this only CSV? or any format? Could we just name this `Save`?

Reply:
All our saves are CSV only. A proper JSON save would be hierarchical. 
- Ship
   - Component (e.g. radar)
      - Damage to specific attribute. 

The current system saves everything as one big struct. When you repair or downgrade, it checks the stock attributes. It works but because everything is in one file, you can't do any unit tests.